phd-tests/framework/src/zfs.rs

```diff
@@ -0,0 +1,321 @@
+/// itself using `zfs destroy`.
+#[derive(Debug)]
+struct Snapshot {
+    /// The name of this snapshot, used to refer to it as a subject of a ZFS
+    /// operation.
+    name: SnapshotName,
+}
+
+impl Snapshot {
+    /// Takes a snapshot of the supplied `dataset`.
+    fn create_from_dataset(dataset: &DatasetName) -> anyhow::Result<Self> {
+        let snapshot_name = format!("{}@phd-{}", dataset.0, Uuid::new_v4());
+        zfs_command("snapshot", &[&snapshot_name])?;
+
+        Ok(Self { name: SnapshotName(snapshot_name) })
+    }
+}
+
+impl Drop for Snapshot {
+    fn drop(&mut self) {
+        debug!(name = self.name.0, "zfs snapshot dropped");
+        let _ = zfs_command("destroy", &[&self.name.0]);
+    }
+}
+
+/// Describes a clone of a specific snapshot. When dropped, attempts to delete
+/// itself using `zfs destroy`.
+#[derive(Debug)]
+struct Clone {
+    /// The name of this clone, used to refer to it as a subject of a ZFS
+    /// operation.
+    name: CloneName,
+
+    /// The point at which this clone is mounted in the global directory
+    /// hierarchy.
+    mount_point: Utf8PathBuf,
+
+    /// The snapshot this clone derives from. Snapshots can't be deleted until
+    /// all their clones are gone; this reference helps to ensure that clones
+    /// and snapshots are deleted in the correct order irrespective of when the
+    /// clones and the [`CloneManager`] are dropped.
+    _snapshot: Snapshot,
+}
+
+impl Drop for Clone {
+    fn drop(&mut self) {
+        debug!(name = self.name.0, "zfs clone dropped");
+        let _ = zfs_command("destroy", &[&self.name.0]);
+    }
+}
+
+/// Represents a specific copy-on-write file within a ZFS clone. When this is
+/// dropped, attempts to delete the associated clone.
+#[derive(Debug)]
+pub struct ClonedFile {
+    /// The clone to which this file belongs.
+    clone: Clone,
+
+    /// The path to this file relative to the mount point of the clone.
+    relative_path: Utf8PathBuf,
+}
+
+impl ClonedFile {
+    /// Creates a snapshot and clone of the dataset that contains the canonical
+    /// location of the file indicated by `path`.
+    pub fn create_from_path(path: &Utf8Path) -> anyhow::Result<Self> {
+        // Canonicalize the path to resolve any symbolic links before doing any
+        // prefix matching.
+        let canonical_path = path.canonicalize_utf8()?;
+
+        let containing_dataset = Dataset::from_path(&canonical_path)
+            .with_context(|| format!("getting dataset containing {path}"))?;
+
+        let relative_file_path = canonical_path
+            .strip_prefix(&containing_dataset.mount_point)
+            .context("getting relative path to file to clone")?;
+
+        let snapshot = Snapshot::create_from_dataset(&containing_dataset.name)?;
+        Self::create_from_paths_and_snapshot(
+            containing_dataset,
+            relative_file_path,
+            snapshot,
+        )
+        .context("creating zfs clone")
```

Comment:
tiny nit: maybe these errors should contain the file path? or do the inner errors already do that?

Reply:
They... do not! Added in fd6ac0b.